[kiatkat] iP

kiatkat opened the pull request:
# PRTS

> Even at time's end, we will meet again.

Part chatbot, part assistant, part productivity tool, and all yours. To all the workaholics who subsist on nothing but instant noodles and your sanity, PRTS is your friend, here to help.

Taking some of that cognitive load off of you, it features
- A text-based interface, for productivity gains
- A small but robust set of functions, for easy pickup
- A distinctive personality, to keep you company

Get it now! Simply
1. Visit this [link](https://www.youtube.com/watch?v=dQw4w9WgXcQ)
2. Profit

Best of all, it's _**free**_! Your wallet won't ache for this one. 🥲

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Here's a preview of a method, `docRest`, in the code:
```java
    public String docRest() {
        if (tasks.isEmpty()) {
            return "How'd you even sleep while standing up... well, have a good rest.";
        } else {
            return "There's still lots of work that needs to be done. " +
                    "We can't afford to have you resting.";
        }
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

kiatkat committed:
Level 0 Implementation - Rename, Greet, Exit

kiatkat committed:
Level 1 Implemented - Echo

kiatkat committed:
Level 2 - Add, List

kiatkat committed:
Added automated testing, minor change to List functionality

kiatkat committed:
Level 3 Implemented. Bugfix for multiple word additions.

kiatkat committed:
Updated testing, minor bugfix

kiatkat committed:
Level 4 Implemented. Minor input validation measures taken and more prompts accommodated.

kiatkat committed:
Level 5 Implemented. Code restructured for future extensibility.

Much of the input validation already existed prior to this commit, but
have been consolidated via the DukeException exception.

kiatkat committed:
Implemented Delete

kiatkat committed:
feat: Implement saving to hard disk

Save the task list to a text file ./data/Duke.txt. If the file or
directory does not exist, creates it.

kiatkat committed:
fix: Handle the case where save file is corrupted

kiatkat committed:
Merge branch 'branch-level-7'

kiatkat committed:
feat: Implement date parsing

kiatkat committed:
feat: Implement delete all

kiatkat committed:
Merge branch 'branch-level-8'

kiatkat committed:
refactor: Rewrite code to use more OOP

The preexisting Duke class was becoming bloated with almost all code
residing within it. Create multiple new classes to extract related
functions. This sets up easier development in the future.

Note: Minor features added - mark and unmark all

kiatkat committed:
Merge branch 'branch-more-oop'

kiatkat committed:
refactor: Divide classes into packages

kiatkat committed:
Merge branch 'origin/add-gradle-support'

kiatkat committed:
build: Set up gradle to build and run Duke

kiatkat committed:
test: Add tests for TaskList

Add two unit tests for the TaskList class, one for addition of a Task
and one for the deletion of the last Task in the TaskList.

kiatkat committed:
test: Add tests for Parser

Add the following tests for Parser, and fix its behaviour to be
consistent with the tests:
- List, correct input
- List, correct input with alternate casing
- List, invalid input

tanshiyu1999 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,153 @@
+public class Parser {
+    public static Command parse(String fullCommand) throws DukeException {
+
+        String[] splits = fullCommand.strip().split(" ", 2);
```
Oh I like that you put the split outside, I split this thing inside which result in a lot of duplicate code.

tanshiyu1999 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,153 @@
+                    throw new DukeException("I can't read your mind. Do add a description.");
+                try {
+                    String[] eventSplit1 = detail.split("/from", 2);
+                    String[] eventSplit2 = eventSplit1[1].split("/to", 2);
```
I think the naming of the variables does not make it very clear exactly what you are splitting into.

kiatkat replied:
Point taken! Thanks for spotting that.

tanshiyu1999 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,54 @@
+
+    Scanner scanner = new Scanner(System.in);
+    static String softbreak = "--------------------------------------------------";
+    static String hardbreak = "——————————————————————————————————————————————————";
```
softbreak and hardbreak to be named as 
SOFTBREAK
HARDBREAK
since they are a constant string variable.

tanshiyu1999 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,153 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeParseException;
+
+public class Parser {
```
Can consider using regex to make your parser more constomizable and less reliant on String manipulation.

tanshiyu1999 reviewed (commented):
LGTM. 

Good project structuring with abstract classes.

kiatkat committed:
refactor: Add new exceptions, Move Parser

DukeException was being used to handle every type of exception, by
passing the error as a string into its constructor. This makes changing
any error messages tedious and complicates exception handling.

Split DukeException into multiple types of exceptions, all inheriting
from DukeException.

Also, move Parser into the command package to prevent wildcard
imports. No other class requires knowledge of all commands.

kiatkat committed:
refactor: Optimize imports and exception handling

Clean up some unused imports, and specified exception types in most
DukeException catches.

kiatkat committed:
Merge branch 'master' into branch-level-9

Major restructuring in master due to new exception handling.

kiatkat committed:
Merge branch 'master' into branch-A-JavaDoc

Major restructuring in master due to new exception handling.

kiatkat committed:
Merge branch 'master' into branch-A-CodingStandard

Major restructuring in master due to new exception handling.

kiatkat committed:
docs: Add Javadoc, Rename Duke

Add Javadoc to all non-private classes and methods.

Rename all instances of Duke to Prts, this has no outward-facing
effects on the user, and no other effects on development.

kiatkat committed:
style: Adjust switch statements

Adjust the swtich statements to follow coding standards

Note: Add some missing modifiers for some fields

kiatkat committed:
feat: Implement find

Add the ability for the user to search for terms in the task
descriptions.

kiatkat committed:
style: Remove wildcard imports

kiatkat committed:
Merge branch 'branch-level-9'

kiatkat committed:
Merge branch 'branch-A-CodingStandard'

Conflicts were not major.

kiatkat committed:
Merge branch 'branch-A-JavaDoc'

Significant conflict due to the renaming of duke throughout the project.
Prts is kept as the new name, and changes from the other branches are
moved over.

MadLamprey commented on src/test/java/prts/ParserTest.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Test
+    public void parse_listMultipleWords_reject() {
+        try {
```
Test case methods follow naming convention!

MadLamprey commented on src/main/java/prts/Prts.java:
```diff
@@ -0,0 +1,65 @@
+                ui.showSoftLine(); // show the divider line ("_______")
+                Command c = CommandParser.parse(fullCommand);
+                c.execute(tasks, ui, storage);
+                isExit = c.isExit();
```
Good division between the uses of each - UI, Parser and Command Class

MadLamprey commented on src/main/java/prts/Prts.java:
```diff
@@ -0,0 +1,65 @@
+    private final Ui ui;
+
+    private String fileName = "prts.PRTS.txt";
+    private String[] directories = {"data"};
```
Why is it being stored as an array?

kiatkat replied:
It's possible that the user might want to create the file in a nested directory somewhere, so I wanted to accept multiple directory names as inputs to enable such a functionality in the future. Though I do agree that as of now, this is indeed a bit redundant as this field is effectively final.

E.g. If `directories = {"deeply", "nested", "data", "folder"};`, I'd like to be able to create the save file in the path `user.dir/deeply/nested/data/folder/`.

MadLamprey commented on src/main/java/prts/Ui.java:
```diff
@@ -0,0 +1,75 @@
+     * Lists the current Tasks stored in the TaskList.
+     * @param tasks The current state of the TaskList.
+     */
+    public void list(TaskList tasks) {
```
Could change the name of method to a more explanatory name

MadLamprey commented on src/main/java/prts/command/ParsingException.java:
```diff
@@ -0,0 +1,66 @@
+ */
+public class ParsingException extends PrtsException {
+
+    public enum ExceptionType {
```
Good use of enums instead of creating multiple sub classes

MadLamprey reviewed (commented):
Good Job!

kiatkat committed:
style: Add checkstyle, Clean up style errors

kiatkat committed:
style: Incorporate PR suggestions

kiatkat committed:
fix: Fix issues with tests

The tests were outdated and did not implement necessary abstract
methods. Add methods to resolve this problem.

kiatkat committed:
Merge branch 'branch-A-CheckStyle'

kiatkat committed:
feat: Add GUI

kiatkat committed:
Merge branch 'branch-level-10'

kiatkat committed:
build: Add CI

Add Continuous Integration with Github Actions.

Note: Text I/O tests are no longer relevant. Remove them.

kiatkat committed:
chore: Add assertions

Add assertions to buffer against null inputs at some points.

kiatkat committed:
chore: Remove reference to text ui test file

kiatkat committed:
chore: Adhere to code quality

Make several changes as follows, to adhere to code quality standards:
- Package GUI elements separately
- Extract PrtsGui class from Prts
- Clarify comments ignoring caught exceptions
- Reduce nesting

kiatkat committed:
Merge pull request #1 from kiatkat/branch-A-CI

Branch-A-CI

kiatkat committed:
Merge branch 'master' into branch-A-Assertions

kiatkat committed:
Merge branch 'master' into branch-A-CodeQuality

kiatkat committed:
Merge pull request #2 from kiatkat/branch-A-Assertions

Branch-A-Assertions

kiatkat committed:
Merge branch 'master' into branch-A-CodeQuality

kiatkat committed:
Merge pull request #3 from kiatkat/branch-A-CodeQuality

Branch-A-CodeQuality

kiatkat committed:
fix: Ensure welcome line displays in GUI

kiatkat committed:
feat: Implement Undo

Add the ability to undo one or more commands entered during the current
session.

kiatkat committed:
Merge branch 'branch-C-Undo'

kiatkat committed:
refactor: Extract methods from CommandParser

CommandParser had a single god method in it and nothing else. Extract
individual portions of the method to increase readability and
extensibility.

kiatkat committed:
fix: Fix various minor bugs

kiatkat committed:
Add representative screenshot

kiatkat committed:
Update README

kiatkat committed:
test: Fix incorrect test outputs

kiatkat committed:
chore: Remove unnecessary import

kiatkat committed:
Move updated README to correct directory
